Rogue-Ape-Studios/Crusader
# ✨ Ability system

## Content

An overhaul of the ability system, it now works based on scriptable objects. Note projectiles are different form all other ability types and need a prefab to work.

## Changes

### Added
`Assets/Animations/Player/AbilityLaser.anim` : Was created. Temp animation for the laser ability.
`Assets/Scripts/Ability/Ability.cs` : Was created. Ability base class.
`Assets/Scripts/Ability/AbilityManager.cs` : Was created. Manages new ability system.
`Assets/Scripts/Ability/AbilitySlot.cs` : Was created. Enum to check what slot an ability is in.
`Assets/Scripts/Ability/CastTarget.cs` : Was created. Enum to check the cast target of an ability.
`Assets/Scripts/Ability/Combat/DamageAbility.cs` : Was created. Base Damage ability script.
`Assets/Scripts/Ability/Combat/LaserBeam.cs` : Was created. Laser ability logic.
`Assets/Scripts/Ability/Combat/LineAreaAbility.cs` : Was created. LineAreaAbility base class.
`Assets/Scripts/Ability/Combat/ProjectileAbility.cs` : Was created. ProjectileAbility base class.
`Assets/Scripts/Ability/Combat/ProjectileCollisionDamage.cs` : Was created. Handles collision checks and damage for projectile abilities (should be added to projectile prefab).
`Assets/Scripts/Ability/Combat/RogueApeStudio.Crusader.Ability.Combat.asmdef` : Was created. asmdf for combat abilities.
`Assets/Scripts/Ability/Combat/ScriptableObjects.meta` : Was created. Folder for ability scriptable objects.
`Assets/Scripts/Ability/Combat/SmiteAbility.cs` : Was created. Logic for smite ability.
`Assets/Scripts/Ability/Combat/SpearAbility.cs` : Was created. Logic for spear ability.
`Assets/Scripts/Ability/Combat/SphereAbility.cs` : Was created. SphereAbility base class.
`Assets/Scripts/Ability/Combat/WaveAbility.cs` : Was created. Wave ability logic.
`Assets/Scripts/Ability/CursorPositionCast.cs` : Was created. Handles directions for rotations.
`Assets/Scripts/Ability/RogueApeStudio.Crusader.Ability.asmdef` : Was created. asmdf for abilities.

### Updated
`Assets/Animations/Player/AbilitySmite.anim` : Was updated. Updated animation events.
`Assets/Animations/Player/AbilitySpear.anim` : Was updated. Updated animation events.
`Assets/Animations/Player/AbilityWave.anim` : Was updated. Updated animation events.
`Assets/Animations/Player/DiveForward.anim` : Was updated. Removed old animation events.
`Assets/Animations/Player/Player.controller` : Was updated. Added a state for laser animation.
`Assets/Prefabs/Player/Player_1.prefab` : Was updated. Added new ability system.
`Assets/Prefabs/VFX/Abilities/HolySpear.prefab` : Was updated / renamed. Renamed and added new ability system.
`Assets/Scripts/Player/Abilities/WaveAbility.cs` : Was updated. Changed used namespace from units.knockback to just units.
`Assets/Scripts/Player/Combat/AttackDamage.cs` : Was updated. Changed used namespace from units.knockback to just units.
`Assets/Scripts/Units/Knockback.cs` : Was updated. Changed used namespace from units.knockback to just units.

## Testing

The feature / Bugfix can be testing by following these steps:

1. pull branch
2. go to scripts/ability/combat/scriptableObject
3. change stats and see the effect.
4. change current ability by dragging in a different scriptable object on the player prefab in the ability manager script under abilities.

Closes #191
